pkg/capabilities/consensus/ocr3/reporting_plugin.go

```diff
@@ -0,0 +1,232 @@
+package consensus
+
+import (
+	"context"
+
+	ocrcommon "github.com/smartcontractkit/libocr/commontypes"
+	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	"google.golang.org/protobuf/proto"
+
+	pbtypes "github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/ocr3/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+var _ ocr3types.ReportingPlugin[any] = (*reportingPlugin)(nil)
+
+type responder interface {
+	response(ctx context.Context, resp response) error
+}
+
+type reportingPlugin struct {
+	batchSize int
+	s         *store
+	r         responder
+
+	// TODO: make this dynamic based on the request
+	aggregator pbtypes.Aggregator
+	encoder    pbtypes.Encoder
+}
+
+func newReportingPlugin(s *store, r responder, batchSize int) *reportingPlugin {
+	return &reportingPlugin{
+		s:         s,
+		r:         r,
+		batchSize: batchSize,
+	}
+}
+
+func (r *reportingPlugin) Query(ctx context.Context, outctx ocr3types.OutcomeContext) (types.Query, error) {
+	batch, err := r.s.firstN(ctx, r.batchSize)
+	if err != nil {
+		return nil, err
+	}
+
+	ids := []*pbtypes.RequestId{}
+	for _, r := range batch {
+		ids = append(ids, &pbtypes.RequestId{
+			RequestId:   r.RequestID,
+			NamespaceId: r.NamespaceID,
+		})
+	}
+
+	return proto.Marshal(&pbtypes.Query{
```

Comment:
In each stage we need to be careful not to exceed max size in bytes from OCR config. For Query maybe it could be guaranteed by config validation?